aten/src/ATen/core/boxing/impl/boxing.h

```diff
@@ -5,75 +5,247 @@
-  // TODO Reuse stack vector instead of allocating?
-  torch::jit::Stack stack;
-  torch::jit::push(stack, std::forward<Args>(args)...);
-
-  (*boxed_kernel_func)(functor, opHandle, &stack);
-
-  TORCH_INTERNAL_ASSERT(stack.size() == 1, "A boxed kernel should only push one return to the stack");
-  return std::move(stack[0]).to<Result>();
-}
+struct BoxedKernelWrapper<Result(Args...), std::enable_if_t<has_quantizer_arg<Args...>::value, void>> {
+  static Result call(KernelFunction::InternalBoxedKernelFunction*, OperatorKernel*, const OperatorHandle&, Args... args) {
+    TORCH_INTERNAL_ASSERT(false, "Unboxed call to a boxed kernel with unboxable parameter type at::Quantizer.");
+  }
+};
 
-// SFINAE version for ops without returns
+// 2. Supported signatures, other than ref-passing.
+//
 template<class Result, class... Args>
-std::enable_if_t<supports_boxing<Result, Args...>::value && std::is_same<void, Result>::value, Result>
-boxAndCallBoxedFunc(KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func, OperatorKernel* functor, const OperatorHandle& opHandle, Args... args) {
-  // TODO Reuse stack vector instead of allocating?
-  torch::jit::Stack stack;
-  torch::jit::push(stack, std::forward<Args>(args)...);
-
-  (*boxed_kernel_func)(functor, opHandle, &stack);
+struct BoxedKernelWrapper<
+  Result(Args...),
+  std::enable_if_t<
+    can_box_all<Args...>::value && can_unbox<Result>::value && !is_tuple_of_tensor_refs<Result>::value,
+    void
+  >
+> {
+  static Result call(
+    KernelFunction::InternalBoxedKernelFunction* boxed_kernel_func,
+    OperatorKernel* functor,
+    const OperatorHandle& opHandle,
+    Args... args
+  ) {
+    // TODO Reuse stack vector instead of allocating?
+    torch::jit::Stack stack;
+    stack.reserve(sizeof...(Args));
+    torch::jit::push(stack, std::forward<Args>(args)...);
+
+    (*boxed_kernel_func)(functor, opHandle, &stack);
+
+    return guts::if_constexpr<!std::is_same<void, Result>::value>(
+      [&] (auto delay_check) {
+        TORCH_INTERNAL_ASSERT(
+          stack.size() == 1,
+          "Boxed kernel was expected to push exactly one return value to the stack."
+        );
+        return delay_check(std::move(stack[0]).to<Result>());
+      },
+      [&] {
+        TORCH_INTERNAL_ASSERT(
+          stack.size() == 0,
+          "Boxed kernel for op with void return type pushed one or more return values to the stack."
+        );
+      }
+    );
+  }
+};
+
+// 3. signatures returning a single reference of the same type as
+// their initial argument.
+//
+// Note that the passed kernels are assumed to be for inplace/outplace ops,
+// and the generated BoxedKernelWrapper specializations will simply return
+// the initial argument.
+//
+template<class Result, class... OtherArgs>
+struct BoxedKernelWrapper<
+  Result(Result, OtherArgs...),
```

Comment:
Suggestion for more clarity: I think you can pattern match directly on `Tensor&` here.

So instead of testing `is_tensor_ref<Result>`, you just have `Tensor&(Tensor&, OtherArgs...)` here. I think the net result should be the same.

Reply:
😍